language/vm/vm_runtime/src/runtime.rs

```diff
@@ -113,19 +114,60 @@ impl<'alloc> VMRuntime<'alloc> {
         txn_block: Vec<Transaction>,
         data_view: &dyn StateView,
     ) -> Vec<TransactionOutput> {
-        let txns = txn_block
-            .into_iter()
-            .map(|txn| {
-                SignedTransaction::try_from(txn)
-                    .expect("All transactions should be user transaction")
-            })
-            .collect();
-        execute_block(
-            txns,
-            &self.code_cache,
-            &self.script_cache,
-            data_view,
-            &self.publishing_option,
-        )
+        let mut result = vec![];
+        let blocks = chunk_block_transactions(txn_block);
+        for block in blocks {
+            match block {
+                TransactionBlock::UserTransaction(txns) => {
+                    result.append(&mut execute_user_transaction_block(
+                        txns,
+                        &self.code_cache,
+                        &self.script_cache,
+                        data_view,
+                        &self.publishing_option,
+                    ))
+                }
+                // TODO: Implement the logic for processing system transactions.
+                TransactionBlock::BlockPrologue(_) => unimplemented!(""),
+                TransactionBlock::WriteSet(_) => unimplemented!(""),
+            }
+        }
+        result
+    }
+}
+
+pub(crate) enum TransactionBlock {
+    UserTransaction(Vec<SignedTransaction>),
+    WriteSet(WriteSet),
+    BlockPrologue(BlockMetadata),
+}
+
+pub(crate) fn chunk_block_transactions(txns: Vec<Transaction>) -> Vec<TransactionBlock> {
```

Comment:
What does doing this "chunking" buy us? I guess I don't see how it's simpler to process than a `Vec<Transaction>`. In addition, `Vec<TransactionBlock>` allows us to represent some nonsense states (e.g., a vec of signed transactions followed by a block prologue).

What about something like

``` 
struct WellFormedBlock(Vec<Transaction>);

impl WellFormedBlock {
  pub fn new(Vec<Transaction>) {
    // assert that the first tx is block metadata
    // assert that if there's a write set, it's in the middle
  }

  pub fn prologue(&self) -> &BlockMetadata { ... }
}
```

and then use this to simplify `execute_block_transactions`?

Reply:
The chunking is trying to get the minimal granularity of execution within the VM. The problem to your proposal is that there can be Blocks that are not well formed within the `Vec<Transaction>`. Typically a well formed transaction will start with a block prologue txn and followed by arbitrary user txn or one writeset txn. But in `Vec<Transaction>` it can start with a number of user transaction, which breaks the definition of a well formed block.

Reply:
I'll keep going on this thread so I can relate to something that is said here.
I have to say I don't mind the code as is, and the grouping of transactions according to some rules, however I have a bunch of questions and something I hope we can step back and think about a bit.
I am not clear that the single API that takes a "random" set of transactions is particularly helpful. I wonder if 2 functions - one for the normal executor path which takes a block and one for the state synchronization - would be a better choice.
In that sense we can imagine the function that takes a block needing a more streamlined check (that the prolog is first, etc.) and we can imagine doing it along the line of what @sblackshear was proposing (I am not sure I would make a new type there, unless there is a need for an API - like Sam seemed to imply with the `prologue` function). In any case that function is a verification on the composition of the transactions in the block.
In that respect I also wonder if we should start imposing restriction on how those blocks have to be composed. For instance one could imagine we require the block prologue to be first always, and we could also impose that any special transaction (today the `WriteSet`, in the future any extension to system transactions) is the **only** transaction in a block (possibly after the prologue). I think that focusing on what a proper block should be is a good way to move forward and start thinking about the problem. Or at least I like the restriction I am thinking about (and the fact there are restrictions).
Then we can think about the state synchronizer and what rules should apply in that case. In that sense I can totally see the API used by the state synchronizer to be along the line of what you are building here. In other words something that has the only purpose of creating correct blocks and call the other API (well not really the other entry point but a common private entry point to both to avoid redundant work).
It seems to me that evolving the code from the clear scenario first can lead to a better understanding of the problem and the needs.
Otherwise this break up is a bit difficult to reason about and to justify given it seems detached from a real problem.
Not sure if this is all just my feeling or what....

Reply:
Really dumb question here: why does the VM need to execute a "well formed" block of transactions. i.e. why can't it start in a state that's mid block, execute some transactions, execute a block metadata execute a few more, end in the middle of a block?

It seems fairly painful that you guys have to reconstruct the block boundaries and if this is really a necessity I'm wondering if it's a symptom of a deeper, cross system impedance mismatch

Reply:
@bmaurer a correction first: this code has nothing to do with the VM strictly speaking and despite the names of things (e.g. runtime.rs). The VM will indeed execute what is told to execute no matter what and without any knowledge of blocks boundaries or anything else.
This code is the layer that sits between the Executor and the VM. In a sense this is the VM driver, for lack of a better name.
Your question though is a good question and I am not sure what the answer to that is. That is also what I was trying to understand and to start a conversation on with my "restriction on blocks".
In other words the question to me is whether there should be a concept of proper/correct block and who should be in charge of verifying it. That may have implications on the VM driver.
There is also the question of whether certain transactions should be tagged in some special way (what today is WriteSet and in the future maybe a more detailed set of transaction types) and whether they should be the only transaction in a block or whether they should obey some rules. And though all of that seems somewhat immaterial and not a VM concern, in reality and being pragmatic, it may simplify things to have certain restrictions. For instance, a transaction that changes configuration flags may require a "restart" of the VM and as such being the only transaction (or last maybe - block prologue aside) in a block may help the cause.
Or at least that is where I think this work was going (and I may be missing something)

Reply:
> Really dumb question here: why does the VM need to execute a "well formed" block of transactions. i.e. why can't it start in a state that's mid block, execute some transactions, execute a block metadata execute a few more, end in the middle of a block?

I was indeed trying to implement as what you suggested. The only reason I was grouping user transactions is because we can execute signature verification in parallel, which will give us a big performance boost. This chunking serves to find the minimal unit of execution without having to change the VM configuration (say WriteSet, or resetting gas schedule, etc) rather than building a well-formed block.

Reply:
hmmm I have to say that I was a bit confused by Sam's first comment (just blaming him ;))
It seems to me that breaking up transactions in case there may be system transactions that require a VM restart is a good thing, we probably need it, do we? 
I am not sure I follow the signature checking aspect of it. Can't we just execute signature checking on all transactions? if some don't need to be checked we can skip them, is chunking helping in any way?
Also chunking here is taking block prologue into account, is it? And it's putting a block prologue in its own "group" and separating user transactions according to that, so I am a bit confused as the why we are doing this. Do we need it?
It felt like we were trying to do some form of well-formed grouping....